### ISSUE TITLE
[WIP]Support remote storage (e.g. Amazon s3) in DALI

### ISSUE THREAD

**Issue by xutianming** [NVIDIA/DALI#1281]
Hi all,

I was using DALI with PyTorch recently and am impressed by its excellent performance.
Currently all of my training data have to be placed on local SSD storage to use DALI. But in production, we have to read from remote storage, such as Amazon S3. 
I want to extend DALI to support remote filesystem. 
Do you think it's possible?
Hope to get some early advise from the developers of DALI.

**Comment by awolant** [comment]
Hi, thanks for the question.
If you are willing to contribute we would be happy to help.

For now, we don't see any technical reason for this to be impossible. Although some of our readers are using random access to images (e.x. `FileReader`) it's not necessary. It's mostly about shuffling patterns after all.

I'm not very familiar with Amazon platform. Let me reach out to some colleagues and get back to you.

**Comment by awolant** [comment]
@xutianming  We discussed this and we do not see any other technical issues, so it's definitely doable feature.
We are very interested in adding it, if you are willing to contribute. It might be useful for many other users.

As you mentioned, the way to approach this would be to write something like `RemoteFilesystemReader` and have Amazon S3 as one of the possible use cases (e.g. GCS being other one).
Could you share some more details about your use case and how you plan to implement this.

@jantonguirao also raised a question about mounting Amazon S3 as a storage and using our existing readers. Is that not enough for you for some reason?

**Comment by xutianming** [comment]
> @xutianming We discussed this and we do not see any other technical issues, so it's definitely doable feature.
> We are very interested in adding it, if you are willing to contribute. It might be useful for many other users.
> 
> As you mentioned, the way to approach this would be to write something like `RemoteFilesystemReader` and have Amazon S3 as one of the possible use cases (e.g. GCS being other one).
> Could you share some more details about your use case and how you plan to implement this.
> 
> @jantonguirao also raised a question about mounting Amazon S3 as a storage and using our existing readers. Is that not enough for you for some reason?

S3-mount is a good choice with nearly no extra development. 
We are maintaining multi-tenant clusters. PyTorch workers are running in dockers and share machines with each other. We found that solutions based on FUSE might get stuck in Linux kernel.

Currently, I am planning to extend `FileStream` to support something like `RemoteFileStream` . I am working on a demo for proof of concept.

**Comment by oleksandrlazariev** [comment]
@xutianming were you able to make it work?

**Comment by xutianming** [comment]
> @xutianming were you able to make it work?

Yes, it works and performs well if you have enough bandwidth between your training cluster and the remote storage.

**Comment by oleksandrlazariev** [comment]
@xutianming how you did it? seems like DALI still not supports it

**Comment by xutianming** [comment]
> @xutianming how you did it? seems like DALI still not supports it

I extended DALI by myself. I plan to make a pull request later.

**Comment by oleksandrlazariev** [comment]
@xutianming I'm sure it would be great for the whole DALI's community!!!

**Comment by gautamkmr** [comment]
@xutianming  Did you add the PR ?

**Comment by xutianming** [comment]
> @xutianming Did you add the PR ?

Sorry for the delay. It's still working in progress

**Comment by zw0610** [comment]
so looking forward to this feature!

**Comment by dianaTanasa** [comment]
Hi. Is there any news on this feature?

**Comment by JanuszL** [comment]
Hi @dianaTanasa,

I'm sorry, but there is none from the DALI team's side.

**Comment by vivekpayasi** [comment]
Is there any ETA on this? Desperately need this.

**Comment by JanuszL** [comment]
Hi All,

You can try out [the `external_source` operator](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/general/data_loading/parallel_external_source.html) and see if the data loading performance implemented that way is sufficient (please check especially `parallel` parameter for the best performance).

**Comment by JigaoLuo** [comment]
Hi everyone,
I noticed that this issue is still open and marked as a WIP. Could anyone provide an update on the current status of S3 support? Is there anyone actively working on it?
Thank you!

**Comment by jantonguirao** [comment]
Hi @JigaoLuo. DALI recently added support for S3 remote storage, though it is kind of experimental for now and not documented.
You can now use `s3://` URIs instead of file paths in our readers, or also instead of file_root in FileReader for autodiscovery. Feel free to share your use-case details if you need any assistance with it.

**Comment by jantonguirao** [comment]
We use the AWS SDK to load from S3 storage to host memory. It is not copied to GPU memory by default, but you can achieve that in your pipeline by explicitly requesting to move the data to GPU.
```
# reading to host memory
read, label = fn.readers.file(file_root="s3://imagenet/val")
# example 1: decoding from CPU to GPU memory
images = fn.decoders.image(read, device='mixed')
# example 2: moving data from CPU to GPU memory
label = label.gpu()
```

**Comment by fversaci** [comment]
Nice to have S3 support! :clap: 
I just tried using the FileReader to read images from an S3 URL, hosted on a local MinIO server and it worked smoothly (DALI v1.38).

However, I ran into an issue when setting `pad_last_batch=True`: i.e., a segmentation fault occurs while reading the last batch.
```
[346c2d31eac6:7973 :0:8022] Caught signal 11 (Segmentation fault: invalid permissions for mapped object at address 0x71f3b8850000)
==== backtrace (tid:   8022) ====
 0 0x0000000000042520 __sigaction()  ???:0
 1 0x00000000001a8067 __nss_database_lookup()  ???:0
 2 0x0000000003e1225b std::_Function_handler<void (int), dali::FileReader::RunImpl(dali::Workspace&)::{lambda(int)#1}>::_M_invoke()  ???:0
 3 0x00000000001d3e76 dali::ThreadPool::ThreadMain()  ???:0
 4 0x00000000000dc253 std::error_code::default_error_condition()  ???:0
 5 0x0000000000094ac3 pthread_condattr_setpshared()  ???:0
 6 0x0000000000126850 __xmknodat()  ???:0
=================================
```
Can you reproduce this error as well? If not, just let me know and I will try to prepare a minimal example.

**Comment by jantonguirao** [comment]
If you can, please provide a reproduction snippet, as it will help us investigate this sooner. Thank you

**Comment by jantonguirao** [comment]
@fversaci A fix have been submitted https://github.com/NVIDIA/DALI/pull/5493

**Comment by fversaci** [comment]
Great, I'll make sure to check it out soon then!
